botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+            execution_env=os.environ.get('AWS_EXECUTION_ENV'),
+            crt_version=crt_version,
+        )
+
+    def set_session_config(
+        self,
+        session_user_agent_name,
+        session_user_agent_version,
+        session_user_agent_extra,
+    ):
+        """
+        Set the user agent configuration values that apply at session level.
+
+        :param user_agent_name: The user agent name configured in the
+            :py:class:`botocore.session.Session` object. For backwards
+            compatibility, this will always be at the beginning of the
+            User-Agent string, together with ``user_agent_version``.
+        :param user_agent_version: The user agent version configured in the
+            :py:class:`botocore.session.Session` object.
+        :param user_agent_extra: The user agent "extra" configured in the
+            :py:class:`botocore.session.Session` object.
+        """
+        self._session_user_agent_name = session_user_agent_name
+        self._session_user_agent_version = session_user_agent_version
+        self._session_user_agent_extra = session_user_agent_extra
+        return self
+
+    def with_client_config(self, client_config):
+        """
+        Create a copy with all original values and client-specific values.
+
+        :type client_config: botocore.config.Config
+        :param client_config: The client configuration object.
+        """
+        cp = copy(self)
+        cp._client_config = client_config
+        return cp
+
+    def to_string(self):
+        """
+        Build User-Agent header string from the object's properties.
+        """
+        if self._client_config and self._client_config.user_agent:
+            return self._build_legacy_ua_string()
+
+        components = [
+            *self._build_sdk_metadata(),
+            RawStringUserAgentComponent('ua/2.0'),
+            *self._build_os_metadata(),
+            *self._build_language_metadata(),
+            *self._build_execution_env_metadata(),
+            *self._build_feature_metadata(),
+            *self._build_config_metadata(),
+            *self._build_app_id(),
+            *self._build_extra(),
+        ]
+        return ' '.join([comp.to_string() for comp in components])
+
+    def _build_sdk_metadata(self):
+        """
+        Build the SDK name and version component of the User-Agent header.
+
+        For backwards-compatibility both session-level and client-level config
+        of custom tool names are honored. If this removes the Botocore
+        information from the start of the string, Botocore's name and version
+        are included as a separate field with "md" prefix.
+        """
+        if (
+            self._session_user_agent_name
+            and self._session_user_agent_version
+            and (
+                self._session_user_agent_name != _USERAGENT_SDK_NAME
+                or self._session_user_agent_version != botocore_version
+            )
```

Comment:
When conditionals start to get more complex, I'm usually in favor of breaking them into discrete variables. Not required, just a note.